File: LTNC/STL_Stack_Queue/TT_HT1.cpp
#include <bits/stdc++.h>
using namespace std;

struct Stack {
    vector<char> a;
    void push(char c){ a.push_back(c); }
    char top(){ return a.back(); }
    void pop(){ a.pop_back(); }
    bool empty(){ return a.empty(); }
};

int prec(char op){
    if(op=='+'||op=='-') return 1;
    if(op=='*'||op=='/') return 2;
    if(op=='^') return 3;
    return 0;
}

int main(){
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    string s,line;
    while(true){
        string tmp;
        if(!getline(cin,tmp)) break;
        s+=tmp;
    }
    if(s.empty()) return 0;
    s.erase(remove_if(s.begin(), s.end(), ::isspace), s.end());
    Stack st;
    string out;
    for(char c: s){
        if(c>='a' && c<='z'){
            out.push_back(c);
        } else if(c=='('){
            st.push(c);
        } else if(c==')'){
            while(!st.empty() && st.top()!='('){
                out.push_back(st.top()); st.pop();
            }
            if(!st.empty() && st.top()=='(') st.pop();
        } else { // operator
            while(!st.empty() && st.top()!='('){
                char t=st.top();
                int pt=prec(t), pc=prec(c);
                if(pt>pc || (pt==pc && c!='^')){ 
                    out.push_back(t); st.pop();
                } else break;
            }
            st.push(c);
        }
    }
    while(!st.empty()){
        out.push_back(st.top()); st.pop();
    }
    cout<<out;
    return 0;
}

File: LTNC/STL_Stack_Queue/TT_HT1.py
'''
Cho đầu vào là một biểu thức dạng trung tố, bao gồm:
- Các toán tử: +, -, *, /, ^
- Các tham số là kí tự từ 'a'->'z'
- Các dấu ngoặc đơn thể hiện thứ tự thực hiện phép tính

Viết chương trình chuyển biểu thức đầu vào thành biểu thức dạng hậu tố

Ví dụ:
Input: (a+(b+c))              (chú ý: đầu vào không ghi a+b+c)
Output: abc++

'''

import sys

class Stack:
    def __init__(self): self.a=[]
    def push(self,x): self.a.append(x)
    def pop(self): return self.a.pop()
    def top(self): return self.a[-1]
    def empty(self): return not self.a

prec = {'+':1,'-':1,'*':2,'/':2,'^':3}

s = sys.stdin.read()
if not s: sys.exit()
expr = ''.join(c for c in s.strip() if not c.isspace())

st = Stack()
out = []

for c in expr:
    if 'a' <= c <= 'z':
        out.append(c)
    elif c == '(':
        st.push(c)
    elif c == ')':
        while not st.empty() and st.top()!='(':
            out.append(st.pop())
        if not st.empty() and st.top()=='(':
            st.pop()
    else:  # operator
        while (not st.empty() and st.top()!='(' and
               (prec[st.top()] > prec[c] or
                (prec[st.top()] == prec[c] and c != '^'))):
            out.append(st.pop())
        st.push(c)

while not st.empty():
    out.append(st.pop())

print(''.join(out)) 